src/attached_probe.cpp

```diff
@@ -37,7 +37,7 @@ namespace bpftrace {
  * Kernel functions that are unsafe to trace are excluded in the Kernel with
  * `notrace`. However, the ones below are not excluded.
  */
-const std::set<std::string> banned_kretprobes = {
+const std::set<std::string> banned_kernel_funcs = {
```

Comment:
That doesn't really work. There are tracepoints as well and kprobe can be inside trace_* handling code.
Also kprobe shouldn't have any problem attaching to these funcs.
fentry need per-cpu recursion counter on bpftrace side

Reply:
The changes in this PR were meant as a small stop gap as we're still discussing how to handle the per-cpu recursion counter discussed here: https://github.com/bpftrace/bpftrace/issues/3144

We obviously don't need to land this if you think this is a step backward but it would have prevented the crash we saw at Meta.

Reply:
It seems like a code churn to land "stop gap" now just to revert it later when #3144 is ready. Not like 3144 is that hard. Just a per-cpu counter for fentry progs. I'd rather remove this kprobe attach restriction, since it seems unnecessary. I'm not sure what was the decision back then when it was added.

Reply:
> Not like 3144 is that hard

It wasn't a question of difficulty. There was some concern over how long we'd have to maintain this code and if there would be a perf impact. @danobi mentioned having a feature probe so we know when we can delete this code but I don't think that exists yet.

Also, just curious, why don't we need this for kprobe/kretprobes?

Reply:
the kernel fix for ABBA is wip. Even when it lands 3144 workaround in bpftrace would have to stay. Feature probing is nice to reduce overheard, but the overhead is tiny anyway. per-cpu array lookups are all inlined now.
kprobe cannot hit this issue, because there is already global per-cpu bpf_prog_active counter that protects all kprobes, but at the same time it makes kprobes unreliable with spurious misses. we cannot make all fentry-s unreliable like kprobes in the kernel, because they're used beyond tracing. At the same time bpftrace can add this check for fentrys, since bpftrace is a tracing tool. People won't be using bpftrace for security enforcement.

Reply:
Got it. Thanks for the explanation! I'll wait to see if anyone else chimes in but if I don't hear anything I'll work on the per-cpu counter for fentry (need it for fexit as well? - doesn't seem so).

Reply:
yes. both fentry/fexit.
btw to reduce overhead of extra checks in 3144 you can combine it with this diff.
Do per-cpu cnt only if fentry/fexit are attached to banned_kernel_funcs[]